### ISSUE TITLE
Support rate limiting

### ISSUE THREAD

**Issue by alexanderkjeldaas** [haskell-servant/servant#266]
Most APIs have some sort of rate-limit.

This rate-limit can be per API call (for example AWS does this), or for the whole API.

It might not be reasonable to encode this in the types directly, but some way of configuring this would make sense, at least as a global rate-limit as wrapping every function is tedious boiler-plate.

**Comment by dmjio** [comment]
I believe this is available in the form of wai middleware, https://hackage.haskell.org/package/wai-middleware-throttle-0.2.0.2/docs/Network-Wai-Middleware-Throttle.html

You should be able to plug this in where servant-server generates an `Application` type, so something like this.

```haskell
main = do
  throttler <- initThrottler
  let f = throttle defaultThrottleSettings throttler
  let app = Servant.Server.serve (Proxy :: Proxy API) handlers
  Warp.run 3000 (f app)
```

Should be configurable via the throttle settings, and would provide global rate-limiting.

**Comment by alpmestan** [comment]
A global or more fine-grained rate-limiting can be implemented as a [WAI middleware](https://hackage.haskell.org/package/wai-3.0.4.0/docs/Network-Wai.html#t:Middleware). As a matter of fact, there's a library on hackage that provides precisely this: [wai-middleware-throttle](https://hackage.haskell.org/package/wai-middleware-throttle).

Applying a middleware `mw :: Middleware` to a servant app is as simple as applying `mw` (which is just a function) to the result of `Servant.Server.serve`, as you can see in [this example](https://github.com/haskell-servant/servant/blob/master/servant-examples/wai-middleware/wai-middleware.hs).

I'm not sure one would get a lot of mileage out of being able to specify rate limiting in API descriptions, as this information is very much server-specific and the other interpretations wouldn't really get anything out of it. If you disagree, I'd be curious to hear about it!

**Comment by alexanderkjeldaas** [comment]
Sorry, this issue should be on servant-client, not the serving part.

**Comment by jkarni** [comment]
@alexanderkjeldaas are the middleware options sufficient? If so, can we close this?

(I also don't quite understand your last comment. Do you mean we should have `servant-client` do leg work for rate-limiting? If so, how would that work?)

**Comment by alexanderkjeldaas** [comment]
Yes that's what I mean.  All useful APIs have rate-limits, so *someone* has
to build this support.

**Comment by ehamberg** [comment]
It would be great to have rate-limiting, but I guess that belongs in the `http-client`/`http-client-tls` libraries that `servant-client` uses?

**Comment by jkarni** [comment]
OK, I see. Given that rate-limiting would show up (a) in the server, (b) in the documentation, and (c) in clients, I see the advantage of having a combinator rather than using just middleware. That said, I don't quite know how we'd support rate-limiting in the client. `http-client` doesn't seem to have support for it.

**Comment by jkarni** [comment]
This will get a lot easier with #327 for `servant-server`.

**Comment by sachs4** [comment]
Hello, I have started to do some basic work on this (https://github.com/sachs4/servant/tree/rate-limit). It still has a lot of warts, but if somebody could take a short look over it and could say if the general direction of this change is good I would try to make it better.

**Comment by jkarni** [comment]
@sachs4 nice. I have a couple of questions or comments:

1) I think our experience has been that adding a single combinator with more configuration options (and then possibly type synonyms) is better than having multiple combinators. Thus, instead of `RateLimitHeader`, `RateLimitIP`, etc., `RateLimit ByHeader`, `RateLimit ByIp`, etc. This saves us a lot of instances, especially in packages that treat the configuration option uniformly.

2) The `HasPathToCombinator` is very interesting, even outside of the rate-limiting context. However, the changes are pretty extensive, and (as far as I can tell) there are questions about whether it'd be witnessing distinctions between APIs that we would like to treat equivalently, and moreover it overlaps somewhat with the functionality `RouterStructure`. It would take a lot of discussion, and quite convincing use cases, for this change to be merged. A much less intrusive approach (which may have issues - I haven't thought it through extensively) might be something like:

```
data RateLimit (limitId :: Symbol) limitType limitAmount
-- e.g. `RateLimit "repo-api" ByIP (100 `Per` Minute)
```

And then any sub-apis under (to the left of `:>` of) a particular `RateLimit` combinator would add to the count named in the combinator. In addition to probably simplifying the implementation, this has the advantage, it seems to me, of allowing join limits over sub-APIs that don't adjoin one another.

3) I'd have to think more about the data structure containing the counts. I'm inclined to think that we should be able to not have contention on `Thingy`s. On the other hand, that might not be a worthwhile effort since, as I understand it, that basically corresponds to an API user. There are some existing libraries that deal directly with rate limiting (e.g. [rate-limit](https://hackage.haskell.org/package/rate-limit-1.1.1/docs/Control-RateLimit.html)) that might be interesting to look at. (I'm not saying that's a better way to go - just an alternative.) For the inner map as well, using a PSQueue instead might be worth thinking about.

Thanks a lot!

**Comment by soenkehahn** [comment]
Hey @sachs4,

thanks for looking into this. I just took a quick look at your PR. Tbh, I wouldn't want to merge this for the following reasons:

- We already have a wai Middleware for rate limiting on the server side. On the client side this should ideally be in `http-client` (and then *maybe* `servant-client` could piggy-back on that). So it feels to me that a rate limiting combinator is only useful in a quite small number of cases.
- I don't want to see the `HasServer` class change for this.
- I'm not sure whether I want to see `Delayed` change for this. This is a much less public interface than `HasServer` though, so it might be ok.
- Is there any way to put this in a separate package? I think this would be much better, because it would put the maintenance burden on people that actually care about the feature.

The last point is problematic, since we don't have a good way of letting people write combinators and just put them in separate packages on github and hackage, due to dependency blow-up. @sachs4: But maybe you can manage?

Sorry, I don't have time to write about this problem in depth atm, but I didn't want to hold back on feedback for this PR.

**Comment by sachs4** [comment]
@jkarni, @soenkehahn thank you for your kind feedback. I've started to fix some of your points. The combinators are now combined into one single. The HasPathToCombinator is gone. The changes to the HasServer class have been reverted. I'm starting to rework @jkarni's 3rd point. While writing some test I ran into this issue https://github.com/CRogers/should-not-typecheck/issues/5, @jkarni do you know any fix for this yet? Regarding servant-docs, would the ability to create charts like this https://dev.twitter.com/rest/public/rate-limits be desirable? Is there any specific Markdown standard the servant-docs package should follow?

**Comment by jkarni** [comment]
Great news!

Re: should-not-typecheck. I don't know any fix, but you can use `doctest` to verify type errors exist and say what you expect. I've been thinking about moving away from `should-not-typecheck` to `doctest` because of that issue.

Re: charts. That would definitely be very cool, but I imagine it'd make sense to separate that out as a later PR?

**Comment by osa1** [comment]
Anyone have any progress on this? Applying `wai-middleware-throttle` to the whole `Application` is not always possible (because we don't want to throttle all endpoints), and I don't want to use a `Raw` endpoint just to be able to throttle. I looked at the links above, but it seems like servant types are quite different now, so they're not easily usable.

**Comment by mbg** [comment]
I have just released [`servant-rate-limit`](https://hackage.haskell.org/package/servant-rate-limit) which allows applying different (or no) rate limiting strategies for different endpoints. For example:

```haskell
import Servant
import Servant.RateLimit

type TestAPI
    = RateLimit (FixedWindow 2 50) (IPAddressPolicy "fixed:") :>
      "fixed-window" :>
      Get '[JSON] String
 :<|> RateLimit (SlidingWindow 2 50) (IPAddressPolicy "sliding:") :>
      "sliding-window" :>
      Get '[JSON] String
 :<|> "unrestricted" :>
      Get '[JSON] String
```

[See the tests for a full code example](https://github.com/mbg/wai-rate-limit/blob/main/servant-rate-limit/test/Spec.hs) and [the README for an explanation](https://github.com/mbg/wai-rate-limit#usage-with-servant). 

This library builds on top of my [`wai-rate-limit`](https://hackage.haskell.org/package/wai-rate-limit) library which should support different backends. Currently [there is only a Redis backend](https://hackage.haskell.org/package/wai-rate-limit-redis). Here's an example for how to initialise this:

```haskell
testApi :: Proxy TestApi
testApi = Proxy

main :: IO ()
main = do
    -- connect to the Redis server and construct a backend for the connection
    backend <- redisBackend <$> checkedConnect defaultConnectInfo

    -- stick the Redis backend into the Servant context so that we can access
    -- it when we try to apply rate limiting policies
    let ctx = backend :. EmptyContext

    -- construct the Servant application using the context
    let app = serveWithContext testApi ctx server
```

It should be relatively straight-forward to implement custom strategies (how to apply rate limiting) and policies (how to identify who to apply them to). For example, a custom policy which uses a username from the request vault as identifier (this could e.g. replace `IPAddressPolicy` in `TestApi`):

```haskell
import qualified Data.Vault.Lazy as V

type UserId = ByteString -- for simplicity

{-# NOINLINE userKey #-}
userKey :: V.Key UserId
userKey = unsafePerformIO newKey

data MyPolicy

instance HasRateLimitPolicy MyPolicy where
    type RateLimitPolicyKey MyPolicy = ByteString

    policyGetIdentifier req =
        fromMaybe (error "expected to have a user id in the vault") $
        V.lookup userKey (vault req)
```
